code/test/upgrade.js

```diff
@@ -0,0 +1,410 @@
+  tokenId,
+  dest,
+  amount,
+  data,
+  testTime = Date.now(),
+  getCurrentState = true
+}) => {
+  // calculate counter from testTime
+  const counter = Math.floor(testTime / INTERVAL)
+  const otp = ONEUtil.genOTP({ seed: walletInfo.seed, counter })
+  // calculate wallets effectiveTime (creation time) from t0
+  const info = await walletInfo.wallet.getInfo()
+  const t0 = new BN(info[3]).toNumber()
+  const walletEffectiveTime = t0 * INTERVAL
+  const index = ONEUtil.timeToIndex({ effectiveTime: walletEffectiveTime, time: testTime })
+  const eotp = await ONE.computeEOTP({ otp, hseed: walletInfo.hseed })
+  let paramsHash
+  let commitParams
+  let revealParams
+  // Process the Operation
+  switch (operationType) {
+    // Format commit and revealParams for FORWARD Tranasction
+    case ONEConstants.OperationType.FORWARD:
+    case ONEConstants.OperationType.SET_RECOVERY_ADDRESS:
+      paramsHash = ONEWallet.computeForwardHash
+      commitParams = { address: dest }
+      revealParams = { operationType, dest }
+      break
+    case ONEConstants.OperationType.COMMAND:
+    case ONEConstants.OperationType.TRACK:
+    case ONEConstants.OperationType.RECOVER_SELECTED_TOKENS:
+      paramsHash = ONEWallet.computeGeneralOperationHash
+      commitParams = { operationType, tokenType, contractAddress, tokenId, dest, amount, data }
+      revealParams = { operationType, tokenType, contractAddress, tokenId, dest, amount, data }
+      break
+    case ONEConstants.OperationType.BACKLINK_ADD:
+    case ONEConstants.OperationType.BACKLINK_DELETE:
+    case ONEConstants.OperationType.BACKLINK_OVERRIDE:
+      paramsHash = ONEWallet.computeDataHash
+      commitParams = { operationType, data }
+      revealParams = { operationType, data }
+      break
+    default:
+      console.log(`Invalid Operation passed`)
+      assert.strictEqual('A Valid Operation', operationType, 'Error invalid operationType passed')
+      return
+  }
+  let { tx, authParams, revealParams: returnedRevealParams } = await TestUtil.commitReveal({
+    Debugger,
+    layers: walletInfo.layers,
+    index,
+    eotp,
+    paramsHash,
+    commitParams,
+    revealParams,
+    wallet: walletInfo.wallet
+  })
+  let currentState
+  if (getCurrentState) { currentState = await TestUtil.getState(walletInfo.wallet) }
+  return { tx, authParams, revealParams: returnedRevealParams, currentState }
+}
+
+contract('ONEWallet', (accounts) => {
+  Logger.debug(`Testing with ${accounts.length} accounts`)
+  Logger.debug(accounts)
+  let snapshotId
+  beforeEach(async function () {
+    await TestUtil.init()
+    snapshotId = await TestUtil.snapshot()
+  })
+  afterEach(async function () {
+    await TestUtil.revert(snapshotId)
+  })
+
+  // === BASIC POSITIVE UPGRADE WALLET ====
+
+  // ====== FORWARD ======
+  // Test forwarding to another wallet
+  // Expected result the wallet will be forwarded to
+  it('UP-BASIC-8 FORWARD: must be able to set forward to another wallet', async () => {
```

Comment:
should also test tracked tokens are forwarded, and future payments and NFTs sent to the old wallet are also automatically forwarded

Reply:
Have added complex forwarding testing 
8-1 Forward Existing
8-2 FORWARD-AUTOMATICALLY-NATIVE
8-3 FORWARD-AUTOMATICALLY-TOKENS

Only tokens tracked are being forwarded which means the following are not forwarded
- ERC20 tokens are not being forwarded automatically (unless manually Tracked)
-  any 721 tokens which were transferred using `transferFrom` instead of `safeTransferFrom` 
- any tokens that were manually untracked

**Recommendation**: Write a utility function to scan transaction history for the wallet an update tokens tracked as part of the upgrade command.